### ISSUE TITLE
farm_map_set_bcscale() prevents the bcscale from being set multiple times.

### ISSUE THREAD

**Issue by mstenta** [farmOS/farmOS#368]
Opening this issue as a follow-up to the PR that was merged and then reverted: https://github.com/farmOS/farmOS/pull/365

We discovered that change caused issues elsewhere (https://github.com/farmOS/farmOS/issues/367), so we reverted the change. Which means the original bug still exists. See @paul121 's description https://github.com/farmOS/farmOS/pull/365

**Comment by mstenta** [comment]
I wonder if using `drupal_static()` and `drupal_static_reset()` would work.

Currently, we are using `static $scale;` - which cannot be unset with `unset($scale);`, hence why @paul121 tried setting `$scale = NULL` instead.

Perhaps we could refactor to use `drupal_static()` and `drupal_static_reset()` instead.

https://api.drupal.org/api/drupal/includes%21bootstrap.inc/function/drupal_static/7.x

https://api.drupal.org/api/drupal/includes%21bootstrap.inc/function/drupal_static_reset/7.x

I'm not sure if that will be any different in practice - it may just `$scale = NULL` internally in the same way.

I really don't feel like I have a good understanding for WHY the change caused https://github.com/farmOS/farmOS/issues/367 in the first place...

We know there's a PHP bug with bcscale persisting across requests (https://bugs.php.net/bug.php?id=78330) ... but we are also resetting it ourselves! So it still seems like it should have worked. There must be more to it...

**Comment by mstenta** [comment]
> I really don't feel like I have a good understanding for WHY the change caused #367 in the first place...

I think I have a hunch...

Imagine two identical requests happening simultaneously in the same process, both setting their own `static` scale, but also modifying the global scale (which we know persists across requests in the same process due to a PHP bug: https://bugs.php.net/bug.php?id=78330).

One request comes in, creates `static $scale` and sets it to the default `0`, then sets global scale to `24`, and starts performing its arithmetic. Meanwhile, a second request comes in, before the first one finishes and resets scale to `0`. The second one creates its own `static $scale`, but sets it to `24`, because that's what the global scale is currently set to (because of the first request that hadn't reset it back yet). Then, presumably the first request finishes first and sets the global scale back to `0` (because that's what it saved in its `static $scale` variable. But, then the second request finishes, and sets the global scale back to `24` (because that's what IT saved).

Thus, we end up with `24` set globally after both requests run... which then feeds into any future requests that occur in the same process, continuing the issue until eventually the process itself ends.

That would explain exactly what we saw in https://github.com/farmOS/farmOS/issues/367, after https://github.com/farmOS/farmOS/pull/365 was merged. The egg quick form was giving me an error MOST mornings that I was collecting eggs... but not all the time. The times that it was working seemed much rarer (anecdotally), which makes sense because they probably occurred at a time when the process was relatively new, and the global scale had not yet been polluted.

But the question remains: why did #365 cause it to start happening? And why wasn't/isn't it happening already??

**Comment by mstenta** [comment]
> Perhaps we could refactor to use `drupal_static()` and `drupal_static_reset()` instead.
> I'm not sure if that will be any different in practice - it may just `$scale = NULL` internally in the same way.

I looked into this, and it *does* work differently... so might be worth a try.

The `drupal_static()` function essentially creates a `static $data = array()` and then adds keyed values to that array. So when `drupal_static_reset()` is used, it actually CAN use `unset()` to remove them. Recall that #365 couldn't use `unset()` because it was just a single static variable, rather than an element in an array, and `unset()` cannot unset static variables.

```
    // Unset the $scale variable so that a new bcscale can later be set.
    // Note that we can't use unset() because this is a static variable.
```

This might work to solve "farm_map_set_bcscale() prevents the bcscale from being set multiple times", but I don't know if/why it would work any different than setting `$scale = NULL` (or still why that causes the issue described in #367). And yet... I feel like we should try it? :-)

**Comment by mstenta** [comment]
In the process of these investigations, I found two instances in farmOS where a function MAY return before `farm_map_reset_bcscale()` is called. I fixed those: https://github.com/farmOS/farmOS/commit/8fc2a36b4ef7ae20d9a14dd84a7465a683054e39

I don't think they were the cause of #367 - because they are both in the `farm_area_generate` module, and in theory they are just defensive returns that shouldn't run in normal conditions.

**Comment by mstenta** [comment]
Ah ha! I think I figured this out.

TL;DR: the `depth--` line is in the wrong place in `farm_map_set_bcscale()`.

I ran some simple tests where I repeatedly called `farm_map_set_bcscale()` and `farm_map_reset_bcscale()`, while printing the result of `bcscale()` after each, to see how it was behaving in different conditions.

This showed two things:

1. `farm_map_set_bcscale()` prevents the bcscale from being set multiple times (confirmed what we already knew)
2. When the change from #365 is applied, and `farm_map_set_bcscale()` + `farm_map_reset_bcscale()` are run more than once, it does NOT reset the scale.

A simple proof that can be run in Devel module's /devel/php:

```
// Ensure that the scale is 0 globally to start.
bcscale(0);

// Set (to 24) and reset (to 0) twice.
farm_map_set_bcscale();
farm_map_reset_bcscale();
farm_map_set_bcscale();
farm_map_reset_bcscale();

// Print the current scale. Should be 0, but it is actually 24.
dpm(bcscale());
```

The expectation is that the final `dpm(bcscale());` statement will output `0`, but in fact it outputs `24`.

This is due to a simple issue with the way `$depth` is used in `farm_map_set_bcscale()`. It is only being decremented when `$depth > 1`, which means that when `farm_map_set_bcscale()` runs a second time, it's starting at `1`, instead of `0` as it should.

`$depth` should be decremented EVERY TIME `farm_map_reset_bcscale();` is run.

So the solution is to simply change:

```
    if ($depth > 1) {
      $depth--;
      return;
    }
```

to:

```
    $depth--;
    if ($depth > 0) {
      return;
    }
```

This, combined with the change from #365, seems to fix all the issues.

Here is a more complex proof that also tests the nested `$depth`:

```
dpm('Test default scale');
dpm('Expectation: 0 / Actual: ' . bcscale());

dpm('Test setting scale');
farm_map_set_bcscale();
dpm('Expectation: 24 / Actual: ' . bcscale());

dpm('Test resetting scale');
farm_map_reset_bcscale();
dpm('Expectation: 0 / Actual: ' . bcscale());

dpm('Test setting+resetting scale in two levels');
farm_map_set_bcscale();
dpm('Expectation: 24 / Actual: ' . bcscale());
farm_map_set_bcscale();
dpm('Expectation: 24 / Actual: ' . bcscale());
farm_map_reset_bcscale();
dpm('Expectation: 0 / Actual: ' . bcscale());
farm_map_reset_bcscale();
dpm('Expectation: 0 / Actual: ' . bcscale());

dpm('Test setting+resetting one more time');
farm_map_set_bcscale();
dpm('Expectation: 24 / Actual: ' . bcscale());
farm_map_reset_bcscale();
dpm('Expectation: 0 / Actual: ' . bcscale());

dpm('Hard reset back to 0');
bcscale(0);
dpm('Expectation: 0 / Actual: ' . bcscale());
``` 

Here is the output WITH the change from #364, but WITHOUT fixing the `$depth` decrementing:

> Test default scale
> Expectation: 0 / Actual: 0
> 
> Test setting scale
> Expectation: 24 / Actual: 24
> 
> Test resetting scale
> Expectation: 0 / Actual: 0
> 
> Test setting+resetting scale in two levels
> Expectation: 24 / Actual: 24
> Expectation: 24 / Actual: 24
> Expectation: 24 / Actual: 24
> Expectation: 0 / Actual: 24
> 
> Test setting+resetting one more time
> Expectation: 24 / Actual: 24
> Expectation: 0 / Actual: 24
> 
> Hard reset back to 0
> Expectation: 0 / Actual: 0

(Notice that the scale is not being reset to 0 until as expected.)

And here is the output with the `$depth--` fix:

> Test default scale
> Expectation: 0 / Actual: 0
> 
> Test setting scale
> Expectation: 24 / Actual: 24
> 
> Test resetting scale
> Expectation: 0 / Actual: 0
> 
> Test setting+resetting scale in two levels
> Expectation: 24 / Actual: 24
> Expectation: 24 / Actual: 24
> Expectation: 24 / Actual: 24
> Expectation: 0 / Actual: 0
> 
> Test setting+resetting one more time
> Expectation: 24 / Actual: 24
> Expectation: 0 / Actual: 0
> 
> Hard reset back to 0
> Expectation: 0 / Actual: 0

What this means is: the `$depth` issue existed already. But, the fact that "farm_map_set_bcscale() prevents the bcscale from being set multiple times" actually PREVENTED the issue from manifesting. So fixing one issue revealed the other.

> Perhaps we could refactor to use `drupal_static()` and `drupal_static_reset()` instead.
> I'm not sure if that will be any different in practice - it may just `$scale = NULL` internally in the same way.

Worth noting: I tested this with `drupal_static()` and `drupal_static_reset()` and it behaves exactly the same as `$static = NULL`.

**Comment by mstenta** [comment]
Created a pull request: https://github.com/farmOS/farmOS/pull/369

**Comment by mstenta** [comment]
Merged #369
